### PR TITLE
klipper: additional config options

### PR DESCRIPTION
**Pull request by cab404**
###### Description of changes

This allows for easier interop with Moonraker, as well as giving an
ability to store klipper configuration files in /var/lib/klipper, thus not
littering /etc with all the backups SAVE_CONFIG does.

~~For backward compatibility, default config location is left as it is, however it would be nice to move it to /var/lib/klipper.~~

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/modules/services/misc/klipper.nix [hunk 1]
```diff
@@ -35,6 +35,30 @@ in
...
+          Path to initial Klipper config. If config at `configPath` doesn't exist, this will be copied it it's place.
+        '';
+      };
+
+      gcodeFolder = mkOption {
+        type = types.nullOr types.path;
+        default = null;
+        example = "/var/lib/klipper/gcodes";
+        description = lib.mdDoc ''
+          Path to Klipper GCode folder. This will create and set permissions on one if it doesn't exist.
```
Comment (lovesegfault) [inline nixos/modules/services/misc/klipper.nix:58]:
```suggestion
          Path to Klipper GCode folder. It will be created if it doesn't already exist.
```

File: nixos/modules/services/misc/klipper.nix [hunk 2]
```diff
@@ -35,6 +35,30 @@ in
...
+        default = "/etc/klipper.cfg";
+        example = "/var/lib/klipper/printer.cfg";
+        description = lib.mdDoc "Path to Klipper config file.";
+      };
+
+      defaultConfig = mkOption {
+        type = types.nullOr types.path;
+        default = null;
+        description = lib.mdDoc ''
+          Path to initial Klipper config. If config at `configPath` doesn't exist, this will be copied it it's place.
```
Comment (lovesegfault) [inline nixos/modules/services/misc/klipper.nix:49]:
```suggestion
          Path to the base Klipper config. If `services.klipper.configPath` is `null`, then this will be copied in its place.
```

File: nixos/modules/services/misc/klipper.nix [hunk 3]
```diff
@@ -35,6 +35,30 @@ in
         description = lib.mdDoc "Path of the API socket to create.";
       };
 
+      configPath = mkOption {
+        type = types.nullOr types.path;
+        default = "/etc/klipper.cfg";
+        example = "/var/lib/klipper/printer.cfg";
+        description = lib.mdDoc "Path to Klipper config file.";
+      };
```
Comment (lovesegfault) [inline nixos/modules/services/misc/klipper.nix:43]: Just gate the default on `stateVersion` and add the info on the release notes.

Comment (cab404) [reply nixos/modules/services/misc/klipper.nix:43]: ty, good idea)

File: nixos/modules/services/misc/klipper.nix (hunk 2 above)
Comment (cab404) [reply nixos/modules/services/misc/klipper.nix:49]: eeeeh, nobody likes oxford commas

**Comment by lovesegfault** [comment]
@GrahamcOfBorg eval

**Comment by cab404** [comment]
@lovesegfault Would it be okay if I disable dynamic user by default?

**Comment by cab404** [comment]
@lovesegfault Can I disable dynamic user by default?

That would be nicer for interop — as we can re-use klipper user for moonraker by default.

(cause moonraker upstream is [doing chgrp/chown on it's files](https://github.com/Arksine/moonraker/issues/311), and that messes with klipper if it's using a different group
)

**Comment by cab404** [comment]
Or maybe I don't understand dynamic users enough, and it's okay to have it dynamic, but with a same name?

**Comment by cab404** [comment]
Oh wow there's a lot more to do on this module than I initially anticipated. I need to somehow befriend `settings` with `configFile` now — or at least make `settings` nullable. I guess having readTOML makes it somewhat easier, but still.

**Comment by cab404** [comment]
Okay, that wasn't that hard to patch.

**Comment by cab404** [comment]
Okay, now it builds. Will try it out soon™

File: nixos/doc/manual/release-notes/rl-2211.section.md [hunk 4]
```diff
@@ -168,6 +168,8 @@ Use `configure.packages` instead.
 
 - The `diamond` package has been update from 0.8.36 to 2.0.15. See the [upstream release notes](https://github.com/bbuchfink/diamond/releases) for more details.
 
+- `services.klipper` will now use `/var/lib/klipper` as its default configuration folder, and will populate `/var/lib/klipper/gcodes` for gcode storage.
+
```
Comment (lovesegfault) [inline nixos/doc/manual/release-notes/rl-2211.section.md:172]: IMO, this needs to be rewritten to mention `stateVersion`, otherwise it's hard to grep for. Here's an example of a good one from 22.05's manual:
```
- For new installations `virtualisation.oci-containers.backend` is now set to `podman` by default.
  If you still want to use Docker on systems where `system.stateVersion` is set to to `"22.05"` set `virtualisation.oci-containers.backend = "docker";`.Old systems with older `stateVersion`s stay with "docker".
```

Comment (cab404) [reply nixos/doc/manual/release-notes/rl-2211.section.md:172]: Oh god, that logic is even more perverted than I expected. I guess I'll just drop the note, and leave the config as it is and where it is if `settings` are specified. 
image

Comment (cab404) [reply nixos/doc/manual/release-notes/rl-2211.section.md:172]: gooood, okay

that removes ability to specify initial config with nix options, which is not good
so I am adding `isConfigMutable`, and explicitly naming options — or it will become undecipherable ;D

**Comment by cab404** [comment]
Okay, I think I got it this time ;D

I removed release notes, since now nothing breaking is happening or is version gated — all of the new behavior is gated by `isConfigMutable`. You can also now specify `configFile`, which is mutually exclusive with `settings`. Either is written used for /etc/klipper.cfg if config is not mutable, and as a default value for `mutableConfigPath` otherwise.

It looks like it generates proper systemd files, although I do need some double-checking.

@lovesegfault pleeeease? <3

File: nixos/modules/services/misc/klipper.nix [hunk 5]
```diff
@@ -156,7 +222,7 @@ in
           (mcu: firmware: pkgs.klipper-flash.override {
             mcu = lib.strings.sanitizeDerivationName mcu;
             klipper-firmware = firmware;
-            flashDevice = cfg.settings."${mcu}".serial;
+            flashDevice = default cfg.firmwares."${mcu}".serial cfg.settings."${mcu}".serial;
```
Comment (lovesegfault) [inline nixos/modules/services/misc/klipper.nix:225]: Why is this change needed?

File: nixos/modules/services/misc/klipper.nix [hunk 6]
```diff
@@ -35,6 +35,38 @@ in
...
 
+      isConfigMutable = mkOption {
+        type = types.bool;
+        default = false;
+        example = true;
+        description = lib.mdDoc ''
+          Whether to copy config to mutable folder instead of using one directly from the store.
+          This will only copy config from NixOS configuration if file at `services.klipper.mutableConfigPath` doesn't exist.
+        '';
+      };
```
[the remark this answers is not in the document]
Comment (cab404) [reply nixos/modules/services/misc/klipper.nix:46]: I found that there are quite a few `isX` options in NixOS, so I used it

File: nixos/modules/services/misc/klipper.nix [hunk 7]
```diff
@@ -35,6 +35,38 @@ in
...
+        description = lib.mdDoc "Path to mutable Klipper config file.";
+      };
+
+      configFile = mkOption {
+        type = types.nullOr types.path;
+        default = null;
+        description = lib.mdDoc ''
+          Path to default Klipper config.
+        '';
+      };
```
[the remark this answers is not in the document]
Comment (cab404) [reply nixos/modules/services/misc/klipper.nix:60]: Yes — people might want to use some other folder. Config file is used as a source file — one that gets copied to `mutableConfigPath` if config is mutable, or to `/etc/klipper.cfg` if it is immutable.

[the remark this answers is not in the document]
Comment (cab404) [reply nixos/modules/services/misc/klipper.nix:60]: There is no need to specify config path if it is immutable IMO

File: nixos/modules/services/misc/klipper.nix (hunk 5 above)
Comment (cab404) [reply nixos/modules/services/misc/klipper.nix:225]: Because otherwise there wouldn't be a way to specify serial port if you use `configFile` instead of `settings `

File: nixos/modules/services/misc/klipper.nix (hunk 7 above)
[the remark this answers is not in the document]
Comment (cab404) [reply nixos/modules/services/misc/klipper.nix:60]: Eh, that makes no sense. configFile is used as an alternative source config, instead of settings. There is no need to specify default now.

File: nixos/modules/services/misc/klipper.nix (hunk 6 above)
[the remark this answers is not in the document]
Comment (SuperSandro2000) [reply nixos/modules/services/misc/klipper.nix:46]: No, there are only a handful of nixos options with .is https://search.nixos.org/options?channel=unstable&from=0&size=50&sort=relevance&type=packages&query=.is

File: nixos/modules/services/misc/klipper.nix [hunk 8]
```diff
@@ -35,6 +35,38 @@ in
         description = lib.mdDoc "Path of the API socket to create.";
       };
 
+      isConfigMutable = mkOption {
+        type = types.bool;
+        default = false;
+        example = true;
+        description = lib.mdDoc ''
+          Whether to copy config to mutable folder instead of using one directly from the store.
+          This will only copy config from NixOS configuration if file at `services.klipper.mutableConfigPath` doesn't exist.
```
Comment (SuperSandro2000) [inline nixos/modules/services/misc/klipper.nix:44]:
```suggestion
          Whether to copy the config to a mutable directory instead of using the one directly from the nix store.
          This will only copy the config if the file at `services.klipper.mutableConfigPath` doesn't exist.
```

File: nixos/modules/services/misc/klipper.nix (hunk 7 above)
[the remark this answers is not in the document]
Comment (SuperSandro2000) [reply nixos/modules/services/misc/klipper.nix:60]: 3 options are quite excessive for this functionality. We can easily condense this into two or one if we wanted.

File: nixos/modules/services/misc/klipper.nix [hunk 9]
```diff
@@ -35,6 +35,38 @@ in
...
+        '';
+      };
+
+      gcodeFolder = mkOption {
+        type = types.nullOr types.path;
+        default = "/var/lib/klipper/gcodes";
+        description = lib.mdDoc ''
+          Path to Klipper GCode folder. It will be created if one doesn't exist.
+        '';
+      };
```
Comment (SuperSandro2000) [inline nixos/modules/services/misc/klipper.nix:68]: I would combine this into mutableConfigPath and make it a directory.

File: nixos/modules/services/misc/klipper.nix [hunk 10]
```diff
@@ -102,12 +138,18 @@ in
         message = "Option klipper.group is not set when a user is specified.";
       }
       {
-        assertion = foldl (a: b: a && b) true (mapAttrsToList (mcu: _: mcu != null -> (hasAttrByPath [ "${mcu}" "serial" ] cfg.settings)) cfg.firmwares);
+        assertion = cfg.settings != null -> foldl (a: b: a && b) true (mapAttrsToList (mcu: _: mcu != null -> (hasAttrByPath [ "${mcu}" "serial" ] cfg.settings)) cfg.firmwares);
         message = "Option klipper.settings.$mcu.serial must be set when klipper.firmware.$mcu is specified";
       }
+      {
+        assertion = (cfg.configFile != null) != (cfg.settings != null);
+        message = "You need to either specify klipper.settings or klipper.defaultConfig.";
```
Comment (SuperSandro2000) [inline nixos/modules/services/misc/klipper.nix:146]: Please use the full config names.

File: nixos/modules/services/misc/klipper.nix [hunk 11]
```diff
@@ -118,14 +160,36 @@ in
...
+              [ -e ${printerConfigPath} ] || {
+                mkdir -p $(dirname ${printerConfigPath})
+                cp ${printerConfigFile} ${printerConfigPath}
+                chmod +w ${printerConfigPath}
+              }
+            ''}
+            ${lib.optionalString (cfg.gcodeFolder != null) ''
+              [ -e ${cfg.gcodeFolder} ] || {
+                mkdir -p ${cfg.gcodeFolder}
+              }
```
Comment (SuperSandro2000) [inline nixos/modules/services/misc/klipper.nix:189]:
```suggestion
              mkdir -p ${cfg.gcodeFolder}
```
``mkdir -p`` should be a noop when the directory exists.

File: nixos/modules/services/misc/klipper.nix (hunk 7 above)
[the remark this answers is not in the document]
Comment (cab404) [reply nixos/modules/services/misc/klipper.nix:60]: Well, we can use `mutableConfigFolder` instead of gcode/config paths and just use default names in it always. 
But I didn't see any reason to do so.

And keep in mind that this PR has two separable functionalities really — one is an ability to specify `configFile` instead of `settings`; the other one is a way to store config in a mutable directory. 
I can technically make two separate PRs for those, but still.

File: nixos/modules/services/misc/klipper.nix (hunk 6 above)
[the remark this answers is not in the document]
Comment (cab404) [reply nixos/modules/services/misc/klipper.nix]: Well, you both are probably right — it really does look closer to `mutableUsers` rather than `isContainer`)

File: nixos/modules/services/misc/klipper.nix [hunk 12]
```diff
@@ -95,19 +123,25 @@ in
...
       }
       {
-        assertion = foldl (a: b: a && b) true (mapAttrsToList (mcu: _: mcu != null -> (hasAttrByPath [ "${mcu}" "serial" ] cfg.settings)) cfg.firmwares);
-        message = "Option klipper.settings.$mcu.serial must be set when klipper.firmware.$mcu is specified";
+        assertion = cfg.settings != null -> foldl (a: b: a && b) true (mapAttrsToList (mcu: _: mcu != null -> (hasAttrByPath [ "${mcu}" "serial" ] cfg.settings)) cfg.firmwares);
+        message = "Option services.klipper.settings.$mcu.serial must be set when settings.klipper.firmware.$mcu is specified";
+      }
+      {
+        assertion = (cfg.configFile != null) != (cfg.settings != null);
+        message = "You need to either specify services.klipper.settings or serivces.klipper.defaultConfig.";
```
Comment (lovesegfault) [inline nixos/modules/services/misc/klipper.nix:138]:
```suggestion
        message = "You need to either specify services.klipper.settings or services.klipper.defaultConfig.";
```

**Review by lovesegfault — APPROVED** [review]

**Comment by cab404** [comment]
Finally finished testing it on the real setup, and applied some fixes. This should be ready to merge @SuperSandro2000 @lovesegfault

**Comment by cab404** [comment]
Here's the setup, just for the record: https://github.com/hackerembassy/infra/blob/master/nodes/jigglypuff/klipper.nix

**Comment by cab404** [comment]
Also, klipper flasher misses some dependencies, namely `dfu-util` and `python3`, and that's something that could be fixed better.

(even after adding dfu-util it didn't seem to work — I remember it being not as obvious last time I've tried it and succeeded)
